runtime/wasm/src/host.rs

```diff
@@ -374,423 +444,258 @@ impl RuntimeHostTrait for RuntimeHost {
-            Err(e) => {
-                return Box::new(future::err(e)) as Box<dyn Future<Item = _, Error = _> + Send>
-            }
-        };
-
-        // Map event handlers to (event handler, event ABI) pairs; fail if there are
-        // handlers that don't exist in the contract ABI
-        let valid_handlers =
-            potential_handlers
+    ) -> DynTryFuture<'a, BlockState> {
+        async move {
+            let data_source_name = &self.data_source_name;
+            let abi_name = &self.data_source_contract_abi.name;
+            let contract = &self.data_source_contract_abi.contract;
+
+            // If there are no matching handlers, fail processing the event
+            let potential_handlers = self.handlers_for_log(&log)?;
+
+            // Map event handlers to (event handler, event ABI) pairs; fail if there are
+            // handlers that don't exist in the contract ABI
+            let valid_handlers = potential_handlers
                 .into_iter()
-                .try_fold(vec![], |mut acc, event_handler| {
+                .map(|event_handler| {
                     // Identify the event ABI in the contract
-                    let event_abi = match util::ethereum::contract_event_with_signature(
-                        &contract,
+                    let event_abi = util::ethereum::contract_event_with_signature(
+                        contract,
                         event_handler.event.as_str(),
-                    ) {
-                        Some(event_abi) => event_abi,
-                        None => {
-                            return Err(format_err!(
-                                "Event with the signature \"{}\" not found in \
-                                 contract \"{}\" of data source \"{}\"",
-                                event_handler.event,
-                                abi_name,
-                                data_source_name,
-                            ))
-                        }
-                    };
-
-                    acc.push((event_handler, event_abi));
-                    Ok(acc)
-                });
-
-        // If there are no valid handlers, fail processing the event
-        let valid_handlers = match valid_handlers {
-            Ok(valid_handlers) => valid_handlers,
-            Err(e) => {
-                return Box::new(future::err(e)) as Box<dyn Future<Item = _, Error = _> + Send>
+                    )
+                    .ok_or_else(|| {
+                        format_err!(
+                            "Event with the signature \"{}\" not found in \
+                                contract \"{}\" of data source \"{}\"",
+                            event_handler.event,
+                            abi_name,
+                            data_source_name,
+                        )
+                    })?;
+                    Ok((event_handler, event_abi))
+                })
+                .collect::<Result<Vec<_>, failure::Error>>()?;
```

Comment:
Interesting, I didn't know this works with `map`. I thought it would result in a `Vec<Result<..., failure::Error>>` instead. I assume there's an implicit conversion going on from that to `Result<Vec<_>, failure::Error>` using `From`?

Reply:
Yeah, this is something very clever from the Rust team. Once you know the trick, it's obvious that it would be done this way. Unfortunately, it is difficult to discover this API.

The trick is that `Result<T: FromIterator, E>` implements `FromIterator`. (More precise syntax [here](https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA%2C%20E%3E%3E)). It does the thing you would hope for and expect by iterating trying to collect each item unless it encounters the first error than bailing. This is very powerful because you can use it on any iterator (not just `map`) and for any other `FromIterator` type - One could collect into a `Result<HashMap>` for example.